fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,266 @@
-        pre_mode = self.module.training
-        self.module.train(mode)
-        if self.module.training:
-            assert not self.need_reduction or pre_mode, "incorrect state transition"
-        else:
-            assert not self.need_reduction, "try to enter eval with grads unreduced"
-        return self
-
-    @contextmanager
+            _ = list(map(lambda x: x.wait(), work_handles))
+
+    @contextlib.contextmanager
     def no_sync(self) -> Generator:
         """A context manager to disable gradient synchronization."""
         old_accumulate_grads = self.accumulate_grads
         self.accumulate_grads = True
         yield
         self.accumulate_grads = old_accumulate_grads
 
-    def forward(self, *inputs: Any, **kwargs: Any) -> Tensor:
-        if self.module.training:
-            if self.need_reduction:
-                raise RuntimeError("OssDdp requires explicit reduction, must call OssDdp.reduce")
-            if not self.accumulate_grads:
-                self.need_reduction = True
-            if self.broadcast_buffers and len(list(self.module.buffers())) > 0:
-                self._sync_buffers()
+    def _find_rank(self, param: Parameter) -> Tuple[OSS, int]:
+        """ Look up where this parameter belongs to """
+        for optim in self.sharded_optimizers:
+            if param in optim.param_to_rank.keys():
+                return optim, optim.param_to_rank[param]
 
-        return self.module(*inputs, **kwargs)
+        assert False, "This parameter is not present in an optimizer, this should not happen"
+        return (None, -1)
 
-    def reduce(self) -> None:
+    def _get_reduce_fn(
+        self, index: int, param: torch.Tensor, should_bucket: bool, dst_rank: int, optimizer: OSS
+    ) -> Callable:
         """
-        This function must be called explicitly after backward to reduce
-        gradients. There is no automatic hook like c10d.
+        Two possible backward hooks for a given parameter: either directly reduce to the appropriate rank,
+        or contribute to a bucket and reduce when the bucket is full.
+
+        Either way a delayed action is necessary and is passed as a callback.
         """
-        assert self.module.training, "Cannot call reduce in eval"
 
-        if not self.need_reduction or self.accumulate_grads:
-            return
+        def gatekeeper() -> None:
+            Variable._execution_engine.queue_callback(optimizer._consume_work_handles)
+
+            # Reset all the grad reduce and bucket state flags
+            self._grad_to_be_reduced = [True] * len(self._grad_to_be_reduced)
+
+        def reduce_direct(*_: Any) -> None:
+            # Skip gradient reduction, do not alter status flags
+            if not self.accumulate_grads and self._grad_to_be_reduced[index]:
+                assert param.grad is not None
+
+                # Make sure that this is not fired twice
+                self._grad_to_be_reduced[index] = False
```

Comment:
What if its a shared parameter (e.g. shared embedding table)?

Reply:
good point, I'm not super clear with that actually, looking into it. How does the BW proceed in that case, the grads are updated twice ?

Reply:
If it's the same param, there is only one param object and one gradient, right? So should be fine. If you look at `model.parameters()` the shared param should only be there once - the parameters and grads are agnostic to the architecture itself.